src/main/kotlin/sh/hsp/techtree/FileSystemYamlReader.kt

```diff
@@ -5,15 +5,65 @@ import com.fasterxml.jackson.dataformat.yaml.YAMLFactory
 import com.fasterxml.jackson.module.kotlin.readValue
 import com.fasterxml.jackson.module.kotlin.registerKotlinModule
 import java.io.File
+import java.io.IOException
 import java.net.URL
 
 class FileSystemYamlReader : YamlReader {
     override fun readToModel(resource: URL): TreeModel {
         val mapper = ObjectMapper(YAMLFactory()).registerKotlinModule()
         val file = File(resource.toURI())
         val yamlData = file.readText()
+        val treeModel: TreeModel = mapper.readValue(yamlData)
 
-        return mapper.readValue(yamlData)
+        validateTreeModel(treeModel)
+
+        return treeModel
+    }
+
+    private fun validateTreeModel(treeModel: TreeModel) {
```

Comment:
Wydzieliłbym walidację do osobnego pliku jako funkcję, która akceptuje `TreeModel`
Według mnie też łamie Single Responsibility Principle, bo `YamlReader` powinien być odpowiedzialny za czytanie yamla, a tu jest odczyt + walidacja 


Wyciągnięcie tego ułatwi testowanie (bo będziesz opierał się na `TreeModel`, a nie konkretnym pliku) + będzie reużywalne dla implementacji `YamlReader` opartego o np. `stdin`

Reply:
To może być też jakaś statyczna funkcja na TreeModelu...
Ogólnie nie jestem pewien, ale Jackson może też wspierać funkcje walidacyjne - ale to nie jest na ten moment.

Reply:
Tak właśnie myślałem, że ta walidacja powinna być bardziej oddzielona. Tylko właśnie nie wiedziałem, gdzie ją umieścić, żeby miała jakiś sens.